### DISCUSSION TITLE
Is there a unique id per case?

### DISCUSSION THREAD

**Discussion by troublesprouter** [freelawproject/courtlistener#3548] [Q&A]
Reporter information is not ideal to identify a case because there is often more than one reporter number per case.

Courlistener ids like `opinion/1900606/association-of-administrative-law-judges-inc-v-heckler/` are not ideal because sometimes there is more than one page, and therefore more than one id or url, for a single legal case.

Is there a one-to-one identifier for cases anywhere?

We need to reliably keep track of which cases we have processed and which we haven't, hence the need for a unique identifier.

Thanks.

**Comment by mlissner** [comment]
Can you elaborate on this, please or provide an example? 

> because sometimes there is more than one page, and therefore more than one id or url, for a single legal case.

**Reply by troublesprouter** [reply]
Thank you.

For example, if you go to 564 U.S. 462, namely https://www.courtlistener.com/c/us/564/462/
you see two results:

https://www.courtlistener.com/opinion/219617/stern-v-marshall/
and
https://www.courtlistener.com/opinion/7345717/stern-v-marshall/

These are different ids but the same opinion.

Conversely, that same case also has reporter numbers 131 S. Ct. 2594 and 180 L. Ed. 2d 475, and therefore all three reporter numbers take you to a single opinion:

 "https://www.courtlistener.com/c/us/564/462",
 "https://www.courtlistener.com/c/sct/131/2594",
 "https://www.courtlistener.com/c/led2d/180/475"

**Reply by flooie** [reply]
Hi @troublesprouter 

I think I can answer your questions.  First you found a duplicate of the same case - which was resulting in both cases showing up with the citation look up tool.  There should be 1 singular opinion_cluster_id - which in this case is `219617`. I axed the newer (larger) cluster from our DB after reviewing your post.  

We've been undergoing a rather large merger or millions of opinions from multiple sources, and as you can imagine we defaulted to creating duplicates over merging incorrect data together.  As we continue to improve our dataset we are finding and fixing these issues.  

Because the case was duplicated - the citation lookup tool took you to the same opinion.  One thing that may be surprising, is that citations are often not unique.  We have thousands if not tens of thousands of examples of opinions sharing the same citation.  This is often because small one line opinions from the 1800's were all added and share the same citation.  So are citation tool in this case does not take you directly to an opinion if there are multiple citations that match an opinion.  

This may seem jarring when you are looking at a SCOTUS opinion, because obviously it's a duplicate.  

Does that help?

**Reply by troublesprouter** [reply]
Yes it does help, thank you @flooie .

I was going to tell you about the non-uniqueness of citations too but couldn't find an example at hand.

Ok, in that case, given what we've been doing, we have found several cases that are duplicated. Would you be interested in the script we are using that has led that to it? Perhaps it could help in identifying the cases that need de-duplication. We could even perhaps adapt and run it ourselves and give you a list of duplicates if it helps. Not sure we can go through the entire database but we can look into it. We're working with a small subset for now.

Regardin unique identifiers, how about this approach: 
- searching by reporter number by going to its url, eg  https://www.courtlistener.com/c/us/564/462/
- if more than one result is present, check the names of the cases with some sort of fuzzy matching to ensure it's not one of the other older cases
- go to the FIRST result on the results page that does match in name, and take that ID as the canonical ID.


Would that be a reliable ID and process to use?

Much appreciated.

**Reply by troublesprouter** [reply]
(if you have an example of URL that takes you to different cases with a single reporter number it would be great to have it. I saw one once but I haven't come up another after, to use it to test the script)

**Reply by flooie** [reply]
@troublesprouter 

I hate to start this with "it depends" but it depends.  What is reliable and what courts are you looking at.  If you are looking at SCOTUS opinions - then you may be safe for the most part.  If you are looking at older opinions or a broader subset of New York courts, I do not think your approach will reliably work 100% of the time. 

First off, sometimes duplicates are caused by updates to an opinion, for example an addendum being added or reargument.  They may not appear as the same opinion but could share a citation.  This may cause issues with some appellate courts but you're probably safe with scotus for the most part.  Outside the ivory tower, not so much.

When opinions are short - say ten or twenty to a page - you often get very similar names... `In re. Orr`, `in re Ott`, or you might get `35 Smith Street vs Mrs. Jones` and `35 Smith Street vs Mr. Jones`.  These are made up - but I've seen similar examples of a wife and husband before or siblings having appeals denied back to back.  These nearly identical opinions are very difficult to disambiguate.  Especially because the text of the opinion can be identical.  

Also - I'm not sure what you mean about single reporter number but if you mean identical citations let me post the first two I can find.

https://www.courtlistener.com/c/ad2d/20/918/
https://www.courtlistener.com/c/ad3d/49/897/

**Reply by troublesprouter** [reply]
That's helpful again @flooie.

What do other platforms do to have unique identifiers for each opinion? Is there a standard? Do people simply use a combination of title + year + court + citation or something like that, or is there an established way to produce a universal identifier for an opinion?

If not how would you go about it roughly?

I'm thinking citation and disambiguation with year and or title perhaps?

**Reply by mlissner** [reply]
We're always happy to learn about and fix data errors. The more you send at a time, the better, but we'll need you to show your work quite thoroughly, so we can audit it for any errors that you might have made (this is a tricky area to deal with).

> What do other platforms do to have unique identifiers for each opinion? Is there a standard? Do people simply use a combination of title + year + court + citation or something like that, or is there an established way to produce a universal identifier for an opinion?

The fact is, it's hard. Citations aren't unique, docket numbers are often a mess, titles vary (and can change!), etc. We take a heuristic approach, but you can see where that got us. I think typically you can be somewhat safe going with court + docket number. That's usually almost enough?

**Reply by troublesprouter** [reply]
Ok, that's helpful context.

Regarding data errors, what we could do if it helps you is you give us a list of citations (perhaps the full list of all citations you have on the database), and we could output the list of URLs of all duplicates, meaning the URLs of all opinons that are duplicated, or at least the url of the search result that contains the duplicate urls for all such cases. That way can easily confirm that the opinion is indeed duplicated by simply going to the URL.

Not sure if we can run it for the entire database but we can try. 

If it helps, just send over the entire set of citations and we'll run it.

**Reply by mlissner** [reply]
Thanks, I think we can do that with a clever query to our DB though. I think the hard part is figuring out which are valid dups and which are not.

**Reply by troublesprouter** [reply]
Noted
